### ISSUE TITLE
Reproducing results from the paper

### ISSUE THREAD

**Issue by johannes-tum** [lixiaoyu2000/Poly-MOT#35]
We very appreciate your tracking approach. It is a great contribution to the community. Unfortunately we cannot fully reproduce your results with the code base of this repository even after multiple attempts. Neither on the validation, nor when submitting to the official NuScenes test server. We greatly appreciate if you help us with that. Here one example with CenterPoint and the validation set: 

1. In the repository it says: “The accuracy with  CenterPoint  in the paper can be reproduced through the parameters above the current  nusc_config.yaml .”
2. `git clone https://github.com/lixiaoyu2000/Poly-MOT.git`
3. We install a new environment: `cd Poly-MOT/;conda env create -f environment.yaml;conda activate polymot `
4. We get the detections in *.json format  from CenterPoint: 
`mkdir -p data/detector/raw_detector/`
https://github.com/tianweiy/CenterPoint/tree/master/configs/nusc
![image](https://github.com/lixiaoyu2000/Poly-MOT/assets/123452326/8c5a121f-3444-4e95-846c-5f5c27f7c749)
5. Since the path in `data/script/first_frame.py`is hard-coded we update the following paths: `dataset_path` and `detector_path`
6. We run `cd data/script; python first_frame.py`. From there we get: 
```======
Loading NuScenes tables for version v1.0-trainval...
23 category,
8 attribute,
4 visibility,
64386 instance,
12 sensor,
10200 calibrated_sensor,
2631083 ego_pose,
68 log,
850 scene,
34149 sample,
2631083 sample_data,
1166187 sample_annotation,
4 map,
Done loading in 43.981 seconds.
======e 
Reverse indexing ...
Done reverse indexing in 10.2 seconds.
======
Parsing ../detector/raw_detector/infos_val_10sweeps_withve...
Extracting first frame token...
100%|█████████████████████████████████████████████████████
write token table to ../utils/first_token_table/trainval/n...
```
7.  Like before we update the path in `PolyMOT/data/script/reorder_detection.py` and run `python reorder_detection.py`. From there we get: 
```
======
Loading NuScenes tables for version v1.0-trainval...
23 category,
8 attribute,
4 visibility,
64386 instance,
12 sensor,
10200 calibrated_sensor,
2631083 ego_pose,
68 log,
850 scene,
34149 sample,
2631083 sample_data,
1166187 sample_annotation,
4 map,
Done loading in 37.357 seconds.
======
Reverse indexing ...
Done reverse indexing in 9.7 seconds.
======
Parsing ../detector/raw_detector/infos_val_10sweeps_withvelo_...
Parsing ../utils/first_token_table/trainval/nusc_first_token...
write order detection file to ../detector/val/val_centerpoint...
```
The output file is: `PolyMOT/data/utils/first_token_table/trainval/nusc_first_token.json`

8. We run the evaluation: 
`cd ~/Poly-MOT; python test.py --nusc_path [...]/nuScenes --config_path config/nusc_config.yaml --detection_path data/detector/val/val_centerpoint.json --first_token_path data/utils/first_token_table/trainval/nusc_first_token.json`
10. Here the results: 
```
### Final results ###
Per-class results:
 AMOTA AMOTP RECALL MOTAR GT MOTA
bicycle 0.571 0.460 0.612 0.815 1993 0.499
bus 0.879 0.492 0.863 0.914 2112 0.789
car 0.857 0.347 0.842 0.867 58317 0.729
motorcy 0.764 0.470 0.765 0.916 1977 0.701
pedestr 0.814 0.374 0.877 0.780 25423 0.678
trailer 0.507 0.989 0.594 0.714 2425 0.424
truck 0.714 0.565 0.750 0.762 9650 0.571
Aggregated results:
AMOTA 0.730
AMOTP 0.528
RECALL 0.758
MOTAR 0.824
GT 14556
MOTA 0.627
MOTP 0.338
MT 4656
ML 1421
FAF 48.8
TP 84358
FP 14040
FN 17237
IDS 302
FRAG 336
TID 0.37
LGD 0.53
Eval time: 3645.8s
```

So that is worse as was reported in the paper: 
![image](https://github.com/lixiaoyu2000/Poly-MOT/assets/123452326/06d5244a-8d1a-4153-b56e-038f3fb43e11)

**Comment by lixiaoyu2000** [comment]
Hi, thank you for your attention. In fact, the overall accuracy of this version of the code is very similar to that in the paper (73.1 AMOTA). The difference in other indicators is because we have reconstructed the original code. We believe that the publicly available code is fully capable of reproducing the accuracy reported in the article, even better after parameter adjustments based on the provided `.yaml`. In addition, parameter fine-tuning is required on the test set.

**Comment by lixiaoyu2000** [comment]
Issue closed due to inactivity.

**Comment by johannes-tum** [comment]
So, this issue is very important for us as we would like to build upon your approach. 
We used the detections officially provided by CenterPoint and LargeKernel without any change and applied your approach with default hyperparameters. 

Here is a comparison of the results: 
| Set | Model | AMOTA | AMOTP | IDS | FP | FN | Source | 
| -------- | ------- | ------- | ------- | ------- | ------- | ------- | ------- |
| Validation | CenterPoint | 73.1 | 52.1 | 281 | - | - | As stated in your paper | 
| Validation | CenterPoint | 73.0 | 52.8 | 302 | - | - | Reproduced using your GitHub code | 
| Validation | LargeKernel | 75.2 | 54.1 | 252 | - | - | As stated in your paper | 
| Validation | LargeKernel | 75.1 | 55.8 | 316 | - | - |Reproduced using your GitHub code | 
| Test | LargeKernel | 75.4 | - | 292  | 19673 | 17956 | As stated in your paper |
| Test | LargeKernel | 75.1 | 55.8 | 316 | 18974 | 19400 | Reproduced using your GitHub code | 

The difference in AMOTA seems small, but the difference in ID Switches and AMOTP is significant.
So did we understand you correctly that we can fully reproduce your results if we change the hyperparameters for the training and test set? Would you be able to share these hyperparameters that you used on the validation set (CenterPoint AND LargeKernel) AND test set?

**Comment by lixiaoyu2000** [comment]
Hi, I really appreciate your interest in Poly-MOT. In response to your problem, I have the following suggestions for your reference:
- The hyperparameters are actually **detector-specific**. For some reasons, we can only provide the centerpoint configuration file (the `nusc_config.yaml` file). However, you can completely reproduce our results whether on centerpoint or LargeKernel (AMOTA can be even better, the sub-metrics may be slightly different) after changing the hyperparameter.
- An optional parameter tuning method is linear search, and we may make this parameter tuning tool public on [Fast-Poly](https://github.com/lixiaoyu2000/FastPoly).
- Your experimental results actually confirm the strong tracking performance of Poly-MOT and its robust ability across detectors under the same parameters. This also proves that Poly-MOT can serve as a powerful baseline. In addition, for new methods, I think ablation experiments are more important than baseline method.

**Comment by johannes-tum** [comment]
What hyperparameters did you change when you submitted to the test server? And which ones for the validation set? The question here refers not to the precise values, but to the names of the parameters.

**Comment by lixiaoyu2000** [comment]
See the implementation details in the article for the parameters listed.

**Comment by lixiaoyu2000** [comment]
Parameters of the same detector are consistent across sets, see experimental details in the article. The core difference between different detectors lies in two core parameters, [association ](https://github.com/lixiaoyu2000/Poly-MOT/blob/main/config/nusc_config.yaml#L25) threshold and [SF ](https://github.com/lixiaoyu2000/Poly-MOT/blob/main/config/nusc_config.yaml#L16) threshold.
